tests/unit/general/test_mdspan.cpp

```diff
@@ -0,0 +1,337 @@
+template<typename T, md_size_t N, typename Layout = MDLayoutLeft<N>>
+using MDArray = MDSpan<mfem::Array<T>, T, N, Layout>;
+
+TEST_CASE("MDArray", "[MDSpan], [MDArray]")
+{
+   SECTION("SetSize")
+   {
+      constexpr int NA = 11, NB = 22, NC = 33;
+      MDArray<int,3> abc;
+      abc.SetSize(NA, NB, NC);
+      REQUIRE(abc.Size() == NA*NB*NC);
+   }
+
+   SECTION("SetLayout")
+   {
+      constexpr int NA = 11, NB = 22, NC = 33;
+      constexpr int na =  0, nb =  1, nc =  2;
+      MDLayout<3> layout_012({na,nb,nc});
+
+      MDArray<int,3> abc(NA, NB, NC);
+      MDArray<int,3, MDLayout<3>> abc_ini(NA, NB, NC);
+      MDArray<int,3, MDLayout<3>> abc_set(NA, NB, NC);
+      abc_set.SetLayout(layout_012);
+
+      REQUIRE(abc_set.Offset(na,nb,nc) == abc.Offset(na,nb,nc));
+      REQUIRE(abc_set.Offset(na,nb,nc) == abc_ini.Offset(na,nb,nc));
+   }
+}
+
+// A possible MDVector class
+template<md_size_t N, typename Layout = MDLayoutLeft<N>>
+struct MDVector : public MDSpan<mfem::Vector, double, N, Layout>
+{
+   using base_t = MDSpan<mfem::Vector,double, N, Layout>;
+
+   MDVector(): base_t() { }
+
+   template <typename... Ts>
+   MDVector(md_size_t n, Ts... args): MDVector(args...) { base_t::Setup(n, args...); }
+};
+
+TEST_CASE("MDVector", "[MDSpan], [MDVector]")
+{
+   SECTION("SetSize")
+   {
+      constexpr int NA = 11, NB = 22, NC = 33;
+      MDVector<3> abc;
+      abc.SetSize(NA, NB, NC);
+      REQUIRE(abc.Size() == NA*NB*NC);
+   }
+
+   SECTION("SetLayout")
+   {
+      constexpr int NA = 11, NB = 22, NC = 33;
+      constexpr int na =  0, nb =  1, nc =  2;
+      MDLayout<3> layout_012({na,nb,nc});
+
+      MDVector<3> abc(NA, NB, NC);
+      MDVector<3, MDLayout<3>> abc_ini(NA, NB, NC);
+      MDVector<3, MDLayout<3>> abc_set(NA, NB, NC);
+      abc_set.SetLayout(layout_012);
+
+      REQUIRE(abc_set.Offset(na,nb,nc) == abc.Offset(na,nb,nc));
+      REQUIRE(abc_set.Offset(na,nb,nc) == abc_ini.Offset(na,nb,nc));
+   }
+
+   SECTION("Offset")
+   {
+      constexpr int NA = 18, NB = 2, NC = 36, ND = 32;
+      // Fortran col major: (N1:18, 2, 36, Nd:32)
+      //                    (    0, 1,  2,     3)
+      // = 0 + 18( 1 + 2( 2 + 36( 3))) = 3978
+      MDVector<4> left(NA,NB,NC,ND); // default layout is LayoutLeft
+      REQUIRE(left.Offset(0,1,2,3) == 3978);
+
+      // C/C++ row major: (N1:18, 2, 36, Nd:32)
+      //                  (    0, 1,  2,     3)
+      // = 3 + 32( 2 + 36( 1 + 2(0))) = 1219
+      // = (((0)*2 + 1) * 36 + 2) * 32 + 3
+      MDVector<4> right(NA, NB, NC, ND);
+      right.SetLayout(MDLayout<4>({3,2,1,0}));
+      REQUIRE(right.Offset(0,1,2,3) == 1219);
+
+      MDVector<4,MDLayoutRight<4>> right4(NA, NB, NC, ND);
+      right4.SetLayout(MDLayoutRight<4>({3,2,1,0}));
+      REQUIRE(right4.Offset(0,1,2,3) == 1219);
+   }
+}
+
+// A possible MDGridFunction class
+template<md_size_t N, class Layout = MDLayoutLeft<N>>
+class MDGridFunction : public MDSpan<mfem::GridFunction,double, N, Layout>
```

Comment:
It seems like MDGridFunction should be accessible in a header of it's own, rather than just in the test case.  Is this not ready for prime time yet?

Reply:
Is the plan to have something like `MDGridFunction` coexist alongside the existing `GridFunction` class, or that `GridFunction` would use multidimensional arrays/views internally?

Reply:
The idea is to allow users to derive from this new class access to multi-dimensional features. Depending on the need, we could also incorporate them internally.